homeassistant/components/recorder/__init__.py

```diff
@@ -313,24 +301,48 @@ def _purge_old_data(self):
 
         purge_before = dt_util.utcnow() - timedelta(days=self.purge_days)
 
-        _LOGGER.info("Purging events created before %s", purge_before)
-        deleted_rows = Session().query(Events).filter(
-            (Events.created < purge_before)).delete(synchronize_session=False)
-        _LOGGER.debug("Deleted %s events", deleted_rows)
+        def _purge_events(session):
+            deleted_rows = session.query(Events) \
+                                  .filter((Events.created < purge_before)) \
+                                  .delete(synchronize_session=False)
+            _LOGGER.debug("Deleted %s events", deleted_rows)
+
+        if self._commit(_purge_events):
+            _LOGGER.info("Purged events created before %s", purge_before)
 
-        _LOGGER.info("Purging states created before %s", purge_before)
-        deleted_rows = Session().query(States).filter(
-            (States.created < purge_before)).delete(synchronize_session=False)
-        _LOGGER.debug("Deleted %s states", deleted_rows)
+        def _purge_states(session):
+            deleted_rows = session.query(States) \
+                                  .filter((States.created < purge_before)) \
+                                  .delete(synchronize_session=False)
+            _LOGGER.debug("Deleted %s states", deleted_rows)
+
+        if self._commit(_purge_states):
+            _LOGGER.info("Purged states created before %s", purge_before)
 
-        Session().commit()
         Session().expire_all()
 
         # Execute sqlite vacuum command to free up space on disk
         if self.engine.driver == 'sqlite':
             _LOGGER.info("Vacuuming SQLite to free space")
             self.engine.execute("VACUUM")
 
+    @staticmethod
+    def _commit(work):
```

Comment:
@rhooper I added the `_commit` function based on your recommendation.

The argument passed to commit can be either a callback or a model that will just be added to the session, so the result is quite compact

Still needs more testing, but `tox -e py34 -- -k recorder` passes for now